### PR TITLE
Slop support for phrase queries

### PR DESCRIPTION
**Pull request by halvorboe**
Implements slop for phrase queries (https://github.com/tantivy-search/tantivy/issues/1068).

Currently the phrase query only supports exact matching. The algorithm used does a set intersection between the position of the last term in the phrase and the positions of all the prior terms in the phrase offset by their position in the phrase.

Looked at the Lucene implementation and the description in "Information Retrieval: Implementing and evaluating search engines". 

I modified the current algorithm slightly to keep track of:
- The start of each candidate phrase. Stored in a separate array.
- The first possible end of the phrase. (searching for `a b .` in `a c b d e` it stores the 0 as the start and 1 as the end). There is never any benefit from picking a later instance of the term. This is stored in what was previously called left.

I'll test the efficiency and try to optimize, but the is no change in complexity.

### CODE REVIEW

**Comment by fulmicoton** [comment]
I did not understand the description. Is this really about slop queries?

Slop queries are about relaxing the constraints of phrase query matching.
For instance, with a slop of 1, the query "my bag" matches on a document that contains "my favorite bag".

**Comment by halvorboe** [comment]
@fulmicoton I'm still working on the description. 

It's the same logic as here right?
https://lucene.apache.org/core/7_4_0/core/org/apache/lucene/search/spans/SpanNearQuery.html
https://github.com/apache/lucene/blob/main/lucene/queries/src/java/org/apache/lucene/queries/spans/SpanNearQuery.java
https://github.com/apache/lucene/blob/main/lucene/queries/src/java/org/apache/lucene/queries/spans/NearSpansOrdered.java

The exact definition I'm working off is:
Given a slop value, a phrase matches the query if the sequence from the phrase appears within `length of the phrase + slop` token positions.

If ~ marks slop:
`a b b c d a c d`
`"c d"` -> 2 matches
`"a d"` -> 0 matches
`"a d"~1` -> 1 matches (`a c d`) 
`"a d"~3` -> 2 matches (`a b b c d`, `a c d`)
`"a d"~100` -> 3 matches (`a b b c d`, `a c d`, `a b b c d a c d`)

The challenges I'm currently thinking about:
1. ~~Keeping track of both where _all_ the phrases ending at the same token started. Exact matching of a phrase does not allow for multiple matches to end on the same token position, while slop matching does. Also if we want to know how much of the slop was _used_ for the match we can use the start and compare it against the end.~~
2. How is the score for a match calculated? In the case above `a c d` should have a higher score than `a b b c d` when both are a match.
3.  ~~What are the edge cases? For example a document with 1 million a b a b and then searching for "a b"~1000000.~~

File: src/query/phrase_query/phrase_weight.rs [hunk 1]
```diff
@@ -77,8 +80,13 @@ impl PhraseWeight {
             similarity_weight,
             fieldnorm_reader,
             self.scoring_enabled,
+            self.slop,
         )))
     }
+
+    pub fn slop(&mut self, slop: u32) {
```
Comment (fulmicoton) [inline src/query/phrase_query/phrase_weight.rs:87]: Rename set_slop

File: src/query/phrase_query/phrase_query.rs [hunk 2]
```diff
@@ -56,9 +57,15 @@ impl PhraseQuery {
         PhraseQuery {
             field,
             phrase_terms: terms,
+            slop: 0,
         }
     }
 
+    /// Slop allowed for the phrase.
+    pub fn slop(&mut self, value: u32) {
```
Comment (fulmicoton) [inline src/query/phrase_query/phrase_query.rs:65]: set_slop

File: src/query/phrase_query/phrase_query.rs [hunk 3]
```diff
@@ -56,9 +57,15 @@ impl PhraseQuery {
         PhraseQuery {
             field,
             phrase_terms: terms,
+            slop: 0,
         }
     }
 
+    /// Slop allowed for the phrase.
```
Comment (fulmicoton) [inline src/query/phrase_query/phrase_query.rs:64]: Please document the behavior, with examples and an example where this is useful.

For instance "java developer"~1 matching "java software developer"

Comment (fulmicoton) [reply src/query/phrase_query/phrase_query.rs:64]: You need to document the "slop keeps the order", and also the slop is "not an overall budget" specification.

In other words
"senior java developer"~1 does match "senior level distributed java software developer". 
The slop is not a budget that is spread amongst the different tokens. It is a maximum quantity that can be applied on every positions

File: src/query/phrase_query/phrase_scorer.rs [hunk 4]
```diff
@@ -130,12 +131,82 @@ fn intersection(left: &mut [u32], right: &[u32]) -> usize {
     count
 }
 
+/// Intersect twos sorted arrays `left` and `right` and outputs the
+/// resulting array in left.
+///
+/// Condition for match is that the value stored in left is less than the value in right,
+/// and that the difference between the value in left_begin is within
```
Comment (fulmicoton) [inline src/query/phrase_query/phrase_scorer.rs:138]: I think your comment was truncated here.

File: src/query/phrase_query/phrase_scorer.rs [hunk 5]
```diff
@@ -130,12 +131,82 @@ fn intersection(left: &mut [u32], right: &[u32]) -> usize {
...
+///
+/// Returns the length of the intersection
+fn intersection_with_distance(
+    left: &mut [u32],
+    left_begin: &mut [u32],
+    right: &mut [u32],
+    max_distance_to_begin: u32,
+) -> usize {
+    // TODO: Improve variable names?
```
Comment (fulmicoton) [inline src/query/phrase_query/phrase_scorer.rs:147]: I think the names are ok.

**Review by fulmicoton — COMMENTED** [review]
Sorry for the late review.

I skimmed through it. Overall it looks good but I did not go into the detail.
Let me ask @PSeitz  to review that in depth.

File: src/query/phrase_query/phrase_scorer.rs [hunk 6]
```diff
@@ -130,12 +131,82 @@ fn intersection(left: &mut [u32], right: &[u32]) -> usize {
...
+    let mut left_i = 0;
+    let mut right_i = 0;
```
Comment (PSeitz) [inline src/query/phrase_query/phrase_scorer.rs:149]: left_index, right_index would be better imo

File: src/query/phrase_query/phrase_scorer.rs [hunk 7]
```diff
@@ -170,6 +242,7 @@ impl<TPostings: Postings> PhraseScorer<TPostings> {
     }
 
     fn phrase_match(&mut self) -> bool {
+        // Need to add support for slop in phrase_count and phrase_exists.
```
Comment (PSeitz) [inline src/query/phrase_query/phrase_scorer.rs:245]: outdated comment?

File: src/query/phrase_query/phrase_scorer.rs [hunk 8]
```diff
@@ -130,12 +131,82 @@ fn intersection(left: &mut [u32], right: &[u32]) -> usize {
...
+/// resulting array in left.
+///
+/// Condition for match is that the value stored in left is less than the value in right,
+/// and that the difference between the value in left_begin is within
+///
+/// Returns the length of the intersection
+fn intersection_with_distance(
+    left: &mut [u32],
+    left_begin: &mut [u32],
+    right: &mut [u32],
```
Comment (PSeitz) [inline src/query/phrase_query/phrase_scorer.rs:144]: would be good to add a comment about the content of left and right

**Comment by edko99** [comment]
Is there any way to help completing this feature and having it merged?

**Comment by halvorboe** [comment]
I'm looking into it this weekend.

**Comment by fulmicoton** [comment]
@halvorboe Ping!

**Comment by halvorboe** [comment]
Sorry, I've not had any capacity. Going back to work on the PR now.

**Comment by fulmicoton** [comment]
Do you need help to land this PR? @halvorboe 

Also if you are interested we have created a discord channel. I think you are one of the oldest user of tantivy. :) 
https://discord.gg/MT27AG5EVE

**Comment by halvorboe** [comment]
Hey @fulmicoton. Sorry again. I'm going to spend a few hours on this

**Pull request closed without merging**